javaslang/src/main/java/javaslang/collection/Vector.java

```diff
@@ -1387,23 +1265,158 @@ public String toString() {
 
-interface VectorModule {
+        private static final VectorTree<?> EMPTY = new VectorTree<>(emptyArray(), 0, 0, 0);
+        @SuppressWarnings("unchecked")
+        static <T> VectorTree<T> emptyTree() { return (VectorTree<T>) EMPTY; }
 
-    final class Combinations {
+        private final Object[] array;
+        private final int offset, length;
+        private final int depthShift;
 
-        static <T> Vector<Vector<T>> apply(Vector<T> elements, int k) {
-            if (k == 0) {
-                return Vector.of(Vector.empty());
+        private VectorTree(Object[] array, int offset, int length, int depthShift) {
+            this.array = array;
+            this.offset = offset;
+            this.length = length;
+            this.depthShift = depthShift;
+
+            assert length() <= treeSize(branchingFactor(), depthShift);
+        }
+
+        static int treeSize(int branchCount, int depthShift) {
+            final int fullBranchSize = 1 << depthShift;
+            return branchCount * fullBranchSize;
+        }
+
+        VectorTree<T> prependLeaf(Object leading) {
+            final int newSize = length() + 1;
+            int offset = this.offset;
+            Object[] array = asArray(leading);
+            int depthShift = this.depthShift;
+            if (length() > 0) {
+                array = this.array;
+                if (offset == 0) {
+                    final Object[] newArray = new Object[branchingFactor()];
+                    newArray[newArray.length - 1] = array;
+                    array = newArray;
+
+                    depthShift += BRANCHING_BASE;
+                    offset = treeSize(branchingFactor() - 1, depthShift);
+                }
+
+                offset -= 1;
+                array = recursiveSet(array, offset, depthShift, leading, 0);
+            }
+
+            return new VectorTree<T>(array, offset, newSize, depthShift);
+        }
+
+        VectorTree<T> appendLeaf(Object trailing) {
+            final int newSize = length() + 1;
+            int depthShift = this.depthShift;
+            Object[] array = asArray(trailing);
+            if (length() > 0) {
+                array = this.array;
+                if ((offset + newSize) > treeSize(branchingFactor(), depthShift)) {
+                    array = asArray(array);
+                    depthShift += BRANCHING_BASE;
+                }
+
+                array = recursiveSet(array, offset + length(), depthShift, trailing, 0);
+            }
+
+            return new VectorTree<T>(array, offset, newSize, depthShift);
+        }
+
+        VectorTree<T> update(int index, T element) {
+            final Object[] root = recursiveSet(array, offset + index, depthShift, element, 0);
+            return new VectorTree<T>(root, offset, length(), depthShift);
+        }
+
+        /** Since the depth of the tree is always small, a non-tailrecursive call is actually faster than pre-calculating the path (is probably unrolled) */
```

Comment:
Thanks, though this iterated twice over the depth.
I managed to unify depth traversal in a single method, hope you like that also :).